0.8 preview

ssorallen opened the pull request:
0.8-preview-merger is a throw away branch for creating a pull request on GitHub for commenting.

This is a port of core-toolbar from 0.5 to 0.8. There are a few outstanding issues:
- Styling in 0.8?
- CSS media queries for adjusting visibility instead of JS-based `core-media-query`?

Commit:
Explicitly import Polymer 0.8, don't import devDeps

Commit:
Update to use Polymer 0.8

* Target all `::host` selectors to the tag name.
* Remove styling for `::content` since that should be left to the
  instantiator.
* Move narrow styling to media query.
* Use 'justified' instead of 'between' for flex justifying since that's
  the CSS value. Translation of between -> justified doesn't seem worth
  it?
* Target most selectors to classes instead of IDs.

Commit:
Remove unnecessary z-index; pointer-events works

Commit:
Prefer CSS over `<br>` for styling

Commit:
Remove docs for non-ported features

* "indent" can be added by the client. It's specific to the width of
  core-icon-button anyway.
* Use `flex` instead of the .fit class. Client can add negative margins
  to accomplish the same effect.

Commit:
Indent all files according to style guide

Commit:
Add missing, currently implied, dev dependencies

* Ignore bower_components/ dir for anyone who installs this via
  `bower install`.

Commit:
Import new 'layout' component for shared styling

Commit:
Add missing bower.son fields

Commit:
Indent according to style guide and update to 2015

No functional changes here, just whitespace and copyright years.

Commit:
Use layout and its class names in the demo

Commit:
Move class names to computed values

* No direct modification of the DOM
* Paste `classNames` implementation

TODO: Where should classNames live?

cdata commented on core-toolbar.css:
```diff
@@ -7,63 +7,86 @@ Code distributed by Google as part of the polymer project is also
 /* narrow layout */
-:host(.core-narrow),
-:host-context(.core-narrow) {
+/*.core-narrow core-toolbar,
```
If this isn't needed, we can just remove it. It will live on in source control :ghost:

ssorallen replied:
:+1:

cdata commented on core-toolbar.css:
```diff
@@ -7,63 +7,86 @@ Code distributed by Google as part of the polymer project is also
   padding: 0;
+}*/
+
+@media (max-width: 639px) {
```
Media queries :cool:

ssorallen replied:
I'm thinking media query break point should live in a shared style repo?

cdata commented on core-toolbar.css:
```diff
@@ -87,20 +111,24 @@ polyfill-next-selector { content: ':host.core-narrow .toolbar-tools, .core-narro
 }
 
+/* ******************************* */
+/* TODO: Should content be styled? */
```
Should it? My guess is that if this is _not_ a paper element, the answer is, "no." Can @frankiefu or @morethanreal chime in on this?

ssorallen replied:
I wanted to remove it but wanted to run it by you and others first. I'm a fan of "no" as well since that's what I expect from using any random custom element: don't touch my light DOM.

frankiefu replied:
I think is important when you put some buttons inside the toolbar those buttons will automatically have the right paddings/margins so they look good.  Also we can consider dropping core-toolbar and just make a material design version and the stylings should follow the [material design spec](http://www.google.com/design/spec/layout/structure.html#structure-app-bar).

cdata commented on core-toolbar.css:
```diff
@@ -87,20 +111,24 @@ polyfill-next-selector { content: ':host.core-narrow .toolbar-tools, .core-narro
 
 /* elements spacing */
-polyfill-next-selector { content: '.toolbar-tools > *'; }
+/*polyfill-next-selector { content: '.toolbar-tools > *'; }
```
Ditto; if this isn't needed, let's just remove it.

ssorallen replied:
:+1:

cdata commented on core-toolbar.html:
```diff
@@ -65,91 +55,135 @@
-        bar.removeAttribute(this.toLayoutAttrName(old));
+    'use strict';
+
+    function classNames(obj) {
```
I remember you talked about making something like this that is more general purpose. Did we end up putting that somewhere, or is this what we have for now?

ssorallen replied:
Yup, I copy/pasted this in a couple places and am going to pull it into `utils` or something similar.

cdata commented on core-toolbar.html:
```diff
@@ -65,91 +55,135 @@
+        classObj[barJustify === 'justified' ?
+            barJustify :
+            barJustify + '-justified'
+          ] = true;
```
This is a very dense one-liner :foggy: maybe break it out into conditional + assignment for the sake of clarity?

ssorallen replied:
Agreed, I'll split it up.

cdata commented on core-toolbar.html:
```diff
@@ -65,91 +55,135 @@
+      return classNames.join(' ');
+    }
+
+    function calcBarClassName(barJustify) {
```
I'm picking up an intention to make this effectively private implementation. I like this style for effectiveness, but less for comprehensibility of the class as a whole. What are your thoughts?

ssorallen replied:
I originally added it to the object passed to `Polymer`, but it didn't reference `this` at all. I remember a discussion about class functions instead of instance functions, and that's what this should be. Do you know where those should live?

cdata replied:
Sadly, no. I tried an experiment with ES6 the other day that was effective. It went something like this:

```javascript
class XElement extends HTMLElement {
  static fooMethod() {
    // ...
  }

  instanceMethod() {
    // ...
    XElement.fooMethod();
    // ...
  }
}

document.registerElement('x-element', XElement);
```

We can probably do the ES5 equivalent, although it will certainly be more verbose that the current strategy of passing an object-literal into `Polymer`.

cdata commented on core-toolbar.html:
```diff
@@ -65,91 +55,135 @@
+        };
+      },
+
+      calcBottomBarClassName: function(bottomJustify) {
```
`calc` > `compute`? :confused:

ssorallen replied:
`compute` is so long, haha. We did decided on `compute` though didn't we. I'll change.

cdata replied:
Yeah, it does feel overly verbose. Bleh. If we agree on an alternative, I'm game.

cdata commented on metadata.html:
```diff
@@ -1,11 +1,11 @@
-    The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
-    Code distributed by Google as part of the polymer project is also
-    subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+  @license
```
I think we agreed to remove `metadata.html` outright from 0.8-preview branches.

ssorallen replied:
Aha didn't know that. I'll nuke from here and others.

cdata commented on test/basic.html:
```diff
@@ -8,79 +8,79 @@
+  <body>
 
-  <core-toolbar></core-toolbar>
+    <core-toolbar></core-toolbar>
```
`test-fixture`?

ssorallen replied:
Sure thing, I didn't touch this yet.

peterwmwong commented on core-toolbar.html:
```diff
@@ -65,91 +55,135 @@
+      computed: {
+        bottomBarClassName: 'calcBottomBarClassName(bottomJustify)',
+        middleBarClassName: 'calcMiddleBarClassName(middleJustify)',
+        topBarClassName: 'calcTopBarClassName(justify)'
```
Since `calcBottomBarClassName`, `calcTopBarClassName`, `calcMiddleBarClassName` methods just call `calcBarClassName`, couldn't these just call `calcBarClassName` directly?  So...
```
  bottomBarClassName: 'calcBarClassName(bottomJustify)',
  middleBarClassName: 'calcBarClassName(middleJustify)',
  topBarClassName:    'calcBarClassName(justify)'
```

ssorallen replied:
Yup great point. I will consolidate.

ssorallen replied:
Your suggestion surfaced a difference in the `computed` feature that I'd like to fix with https://github.com/Polymer/polymer/pull/1243. Once that or something similar is merged, I can collapse these properties into the same method.

Commit:
Remove "narrow" styles in favor of media queries

Commit:
Remove unneeded styling and comment remaining

`.toolbar-tools` disables touch events so multiple toolbars can stack
and not absorb events. All children must have pointer events re-enabled
to work as expected.

Commit:
Split assignment in two for clarity var name

Commit:
Remove deprecated metadata.html, gone in 0.8

Commit:
Use common `compute` naming for computed props

This was decided in the style guide.

ssorallen commented:
Closing as the initial state of 0.8. Near future work:

* Potentially consolidate on a single compute method if https://github.com/Polymer/polymer/pull/1243 is merged
* Add tests
* Move `classNames` function to a common repo
